src/lib/shell/streamer_esp32.cpp

```diff
@@ -44,6 +44,74 @@ static int chip_command_handler(int argc, char ** argv)
     return static_cast<int>(err.AsInteger());
 }
 
+void get_command_completion(const char * buf, linenoiseCompletions * lc)
+{
+    size_t len = strlen(buf);
+
+    // If there is a trailing space, remove it for completion lookup
+    bool trailing_space = false;
+    if (buf[len - 1] == ' ')
+    {
+        trailing_space = true;
+        memset((char *) &buf[len - 1], 0, 1);
+        len -= 1;
+    }
+
+    const char * matter_prefix = "matter";
+
+    // the commands are prefixed with "matter"
+    if (len >= 6 && strncmp(buf, matter_prefix, 6) == 0)
+    {
+        // remove "matter " prefix for completion lookup as the command
+        // registeration in general are done without "matter" prefix
+        char * line = new char[len - strlen(matter_prefix) + trailing_space + 1];
+        strcpy((char *) line, &buf[7]);
+
+        cmd_completion_context context = cmd_completion_context(line);
+        CHIP_ERROR err_code            = Engine::GetCommandCompletions(&context);
+        if (err_code == CHIP_NO_ERROR)
+        {
+            for (size_t i = 0; i < context.cmdc; i++)
+            {
+
+                // example: "matter" + " " + "config" + " " + "productid" + " " + '\0'
+                char * cmd_completion =
+                    new char[strlen(matter_prefix) + 1 + strlen(context.ret_prefix) + 1 + strlen(context.cmdv[i]->cmd_name) + 1];
+
+                // Write "matter "
+                size_t pos = 0;
+                strncpy(&cmd_completion[pos], matter_prefix, strlen(matter_prefix) + 1);
```

Comment:
strncpy should take the target destination as size or not take any size at all. Use strcpy if you cannot use the size of destination, since strcpy is supposed to protect against destination overflow.

Reply:
In hindsight after seeing all this code: you should probably be using `src/lib/support/StringBuilder.h` to have these append logic work and then you do not need to worry as much about pos and sizes and such.
StringBuilderBase with a ScopedBuffer for memory allocation should work well.

Reply:
thank you very much for taking time to review and the guidance, really appreciate it. Let me work on the changes to address them and will send this again later.